pallets/merkle/src/utils/hasher.rs

```diff
@@ -1,5 +1,249 @@
+	VerificationFailed,
+	InvalidProof,
+}
+
+#[cfg_attr(feature = "std", derive(Debug))]
+#[derive(Clone, Encode, Decode, PartialEq)]
+pub struct Setup {
+	hasher: HashFunction,
+	backend: Backend,
+}
+
+impl Setup {
+	pub fn new(hasher: HashFunction, backend: Backend) -> Self {
+		Self { hasher, backend }
+	}
+
+	pub fn hash(&self, xl: &ScalarBytes, xr: &ScalarBytes) -> ScalarBytes {
+		match self.backend {
+			Backend::Bulletproofs => match self.hasher {
+				HashFunction::PoseidonDefault | HashFunction::Poseidon(6, 3) => {
+					let sl = Scalar::from_bytes_mod_order(slice_to_bytes_32(xl));
+					let sr = Scalar::from_bytes_mod_order(slice_to_bytes_32(xr));
+					Poseidon_hash_2(sl, sr, &DEFAULT_POSEIDON_HASHER).to_bytes().to_vec()
+				}
+				_ => panic!("Unsupported hash function"),
+			},
+			_ => panic!("Unsupported backend"),
+		}
+	}
+
+	pub fn get_bulletproofs_poseidon(&self) -> &Poseidon {
+		match self.backend {
+			Backend::Bulletproofs => match self.hasher {
+				HashFunction::PoseidonDefault => &DEFAULT_POSEIDON_HASHER,
+				_ => panic!("Hasher is not default poseidon hasher"),
+			},
+			_ => panic!("Backend not bulletproofs"),
+		}
+	}
+
+	pub fn generate_zero_tree(&self, depth: usize) -> (Vec<ScalarBytes>, ScalarBytes) {
+		let zero_tree = match self.hasher {
+			HashFunction::PoseidonDefault => {
+				gen_zero_tree(DEFAULT_POSEIDON_HASHER.width, &DEFAULT_POSEIDON_HASHER.sbox)
+			}
+			_ => gen_zero_tree(DEFAULT_POSEIDON_HASHER.width, &DEFAULT_POSEIDON_HASHER.sbox),
+		};
+		(
+			zero_tree[0..depth].iter().map(|x| x.to_vec()).collect(),
+			zero_tree[depth].to_vec(),
+		)
+	}
+
+	pub fn verify_zk(
+		&self,
+		depth: usize,
+		cached_root: ScalarBytes,
+		comms: Vec<ScalarBytes>,
+		nullifier_hash: ScalarBytes,
+		proof_bytes: Vec<u8>,
+		leaf_index_commitments: Vec<ScalarBytes>,
+		proof_commitments: Vec<ScalarBytes>,
+		recipient: ScalarBytes,
+		relayer: ScalarBytes,
+	) -> Result<(), SetupError> {
+		match self.backend {
+			Backend::Bulletproofs => {
+				let cached_root_s = Scalar::from_bytes_mod_order(slice_to_bytes_32(&cached_root));
+				let comms_c = comms.iter().map(|x| CompressedRistretto::from_slice(x)).collect();
+				let nullifier_hash_s = Scalar::from_bytes_mod_order(slice_to_bytes_32(&nullifier_hash));
+				let leaf_index_commitments_c = leaf_index_commitments
+					.iter()
+					.map(|x| CompressedRistretto::from_slice(x))
+					.collect();
+				let proof_commitments_c = proof_commitments
+					.iter()
+					.map(|x| CompressedRistretto::from_slice(x))
+					.collect();
+				let recipient_s = Scalar::from_bytes_mod_order(slice_to_bytes_32(&recipient));
+				let relayer_s = Scalar::from_bytes_mod_order(slice_to_bytes_32(&relayer));
+				self.verify_bulletproofs_poseidon(
+					depth,
+					cached_root_s,
+					comms_c,
+					nullifier_hash_s,
+					proof_bytes,
+					leaf_index_commitments_c,
+					proof_commitments_c,
+					recipient_s,
+					relayer_s,
+				)
+			}
+			_ => panic!("Unsupported backend"),
```

Comment:
Change this to proper `Unimplemented` error.

Reply:
Ok, will do it in https://github.com/webb-tools/anon/pull/144 since this branch is focused on arkworks  stuff

Reply:
ah i just merged that PR, didn't see this until now. It can all be handled separately later too. No rush.